libafl_libfuzzer/build.rs

```diff
@@ -69,20 +74,146 @@ fn main() {
     let mut lib_path = custom_lib_dir.join(std::env::var_os("TARGET").unwrap());
     lib_path.push("release");
 
-    #[cfg(all(feature = "embed-runtime", target_family = "unix"))]
-    {
-        // NOTE: lib, .a are added always on unix-like systems as described in:
-        // https://gist.github.com/novafacing/1389cbb2f0a362d7eb103e67b4468e2b
+    if cfg!(target_family = "unix") {
+        use std::path::Path;
+
+        lib_path.push("libafl_libfuzzer_runtime.a");
+        let target_libdir = Command::new("rustc")
+            .args(["--print", "target-libdir"])
+            .output()
+            .expect("Couldn't find rustc's target-libdir");
+        let target_libdir = String::from_utf8(target_libdir.stdout).unwrap();
+        let target_libdir = Path::new(target_libdir.trim());
+
+        let rust_lld = target_libdir.join("../bin/rust-lld");
+        let rust_ar = target_libdir.join("../bin/llvm-ar"); // NOTE: depends on llvm-tools
+        let rust_objcopy = target_libdir.join("../bin/llvm-objcopy"); // NOTE: depends on llvm-tools
+        let nm = "nm"; // NOTE: we use system nm here because llvm-nm doesn't respect the encoding?
+
+        let redefined_symbols = custom_lib_dir.join("redefs.txt");
+
+        let objfile_orig = custom_lib_dir.join("libFuzzer.o");
+        let objfile_dest = custom_lib_dir.join("libFuzzer-mimalloc.o");
+
+        let mut command = Command::new(rust_lld);
+        command
+            .args(["-flavor", "gnu"])
+            .arg("-r")
+            .arg("--whole-archive")
+            .arg(lib_path)
+            .args(["-o", objfile_orig.to_str().expect("Invalid path characters present in your current directory prevent us from linking to the runtime")]);
+
+        assert!(
+            !command.status().map(|s| !s.success()).unwrap_or(true),
+            "Couldn't link runtime crate! Do you have the llvm-tools component installed?"
+        );
+
+        let mut child = Command::new(nm)
+            .arg(&objfile_orig)
+            .stdout(Stdio::piped())
+            .spawn()
+            .unwrap();
+
+        let mut redefinitions_file = BufWriter::new(File::create(&redefined_symbols).unwrap());
+
+        const NAMESPACE: &str = "🐇";
+        const NAMESPACE_LEN: usize = NAMESPACE.as_bytes().len();
+
+        let replacement = format!("_ZN{NAMESPACE_LEN}{NAMESPACE}");
+
+        // redefine all the rust-mangled symbols we can
+        // TODO this will break when v0 mangling is stabilised
+        for line in BufReader::new(child.stdout.take().unwrap()).lines() {
```

Comment:
can't we just avoid to export symbols when generating the runtime lib? like partial linking and then remove all symbols

Reply:
We want to retain these symbols for both debugging purposes and because we need to retain certain unmangled symbols (e.g. LLVMFuzzerRuntime, LLVMFuzzerMutate, etc.)